### PR TITLE
Infer - and ! using std::ops::{Neg, Not}

### PR DESCRIPTION
**Pull request by kiljacken**
Found some low hanging fruit while taking a cursory look at index inferring.

### CODE REVIEW

**Comment by kiljacken** [comment]
I was unsure whether to completely replace the existing hard coded version for built ins, but it should be a trivial change after this. Assuming Rust behavior for - and ! on built-ins isn't likely to change, it's probably useful to keep it around as a "fast path".

File: crates/ra_hir_expand/src/name.rs [hunk 1]
```diff
@@ -152,6 +152,8 @@ pub const RANGE_INCLUSIVE_TYPE: Name = Name::new_inline_ascii(b"RangeInclusive")
 pub const RANGE_TO_INCLUSIVE_TYPE: Name = Name::new_inline_ascii(b"RangeToInclusive");
 pub const RANGE_TO_TYPE: Name = Name::new_inline_ascii(b"RangeTo");
 pub const RANGE_TYPE: Name = Name::new_inline_ascii(b"Range");
+pub const NEG_TYPE: Name = Name::new_inline_ascii(b"Neg");
+pub const NOT_TYPE: Name = Name::new_inline_ascii(b"Not");
```
Comment (matklad) [inline crates/ra_hir_expand/src/name.rs]: These should be `_TRAIT`, not  `_TYPE`.

**Comment by matklad** [comment]
LGTM, but let's r? @flodiebold

File: crates/ra_hir_expand/src/name.rs (hunk 1 above)
Comment (kiljacken) [reply crates/ra_hir_expand/src/name.rs]: Both Future and IntoIterator are _TYPE so was just following the existing naming. Can still change it if neede of course :)

Comment (matklad) [reply crates/ra_hir_expand/src/name.rs]: Ah, ok, so `FUTURE` is `future` and `FUTURE_TYPE` is `Future`. But `DEFAULT_TRAIT` is trait. I guess it's better to leave as is for the time being, and then just add an `N!` macro, such that `N![Not]` is unambigious.

**Comment by flodiebold** [comment]
Yeah, I think we should keep the builtin implementations at least as long as there are still problems with normalizing associated types in Chalk.

LGTM!
bors r+

**Pull request merged**